pkg/spec3/spec.go

```diff
@@ -48,3 +48,30 @@ func (o *OpenAPI) UnmarshalJSON(data []byte) error {
 	}
 	return json.Unmarshal(data, &p)
 }
+
+func (o *OpenAPI) MarshalJSON() ([]byte, error) {
+	if internal.UseOptimizedJSONMarshalingV3 {
+		return internal.DeterministicMarshal(o)
+	}
+	type OpenAPIWithNoFunctions OpenAPI
+	p := (*OpenAPIWithNoFunctions)(o)
+	return json.Marshal(&p)
+}
+
+func (o *OpenAPI) MarshalNextJSON(opts jsonv2.MarshalOptions, enc *jsonv2.Encoder) error {
+	var x struct {
+		Version      string                 `json:"openapi"`
+		Info         *spec.Info             `json:"info"`
+		Paths        *Paths                 `json:"paths,omitzero"`
+		Servers      []*Server              `json:"servers,omitempty"`
+		Components   *Components            `json:"components,omitzero"`
+		ExternalDocs *ExternalDocumentation `json:"externalDocs,omitzero"`
+	}
+	x.Version = o.Version
+	x.Info = o.Info
+	x.Paths = o.Paths
+	x.Servers = o.Servers
+	x.Components = o.Components
+	x.ExternalDocs = o.ExternalDocs
```

Comment:
Is the only difference between this struct and `OpenAPI` struct the `omitzero` properties?

If that is the case I think it would be possible to avoid a copy here by casting `*OpenAPI` to `*OpenAPIOmitZero`:

Following toy example seemed to compile at least:
```go

type X struct {
	A string
	B string
}

type Y struct {
	A string
	B string
}

func (x X) ConvertToY() Y {
	return Y(x)
}

func (x *X) ConvertToYPointer() *Y {
	return (*Y)(x)
}
```

Reply:
isn't the conversion making a copy anyway?

Reply:
It sounds like you're suggesting doing:
```suggestion
	type X struct {
		Version      string                 `json:"openapi"`
		Info         *spec.Info             `json:"info"`
		Paths        *Paths                 `json:"paths,omitzero"`
		Servers      []*Server              `json:"servers,omitempty"`
		Components   *Components            `json:"components,omitzero"`
		ExternalDocs *ExternalDocumentation `json:"externalDocs,omitzero"`
	}
	x := X(o)
```

Reply:
Yes, except `x = (*X)(&o)`

Reply:
it's definitely simpler.